source/base/coreutils/dircolors.1
.TH DIRCOLORS 1L \" -*- nroff -*-
.SH NAME
dircolors \- Set color parameters for 
.BR ls (1)
.SH SYNOPSIS
.TP
\fBeval `dircolors\fR [\-abckstzPS] [filename]\fB`\fR
.TP
.B dircolors \-v
.SH DESCRIPTION
.B dircolors
is a program to generate a setup for the color version of GNU 
.BR ls .
The setup consists of an
.B LS_COLORS
environment variable and a set of aliases.  For
.B sh
and
.BR ash ,
which lack aliases, shell functions are substituted.  The
.B dircolors
command will generally be run from the user's
.IR .profile ,
.I .cshrc
or equivalent file, or the system-wide equivalents.
.SS OPTIONS
.TP
.I "\-a, \-s"
Assume the user is using a Bourne-style shell which does not support
aliasing.  This is default if the base name of the environment
variable
.B SHELL
is
.B sh
or
.BR ash .
Instead a shell function is used.  If the
.I \-P
option is also used, this function will need to spawn a subshell; this
is slow and should be avoided if possible.
.TP
.I "\-b, \-k"
Assume the user is using a Bourne-style shell that supports Korn-style
aliasing.  This is the default if the base name of the environment
variable
.B SHELL
is
.B bash
or
.BR ksh .
You may want to specify this option explicitly if your
.B sh
is really a more advanced shell, which does support Korn-style aliasing.
.TP
.I "\-z"
Assume the user is using a Bourne-style shell that supports Korn-style
aliasing, but differs between string and list environment variables.
This is the default is the base name of the environment variable
.B SHELL
is
.BR zsh .
.TP
.I "\-c, \-t"
Assume the user is using a C-style shell.  This is the default if the
base name of the environment variable
.B SHELL
is
.B csh
or
.BR tcsh .
.TP
.I "\-P"
Do not seek the path to find the location of the
.B ls
binary.  By default,
.B dircolors
will find the location of the
.B ls
binary at the time it is run, and include it in the shell function or
alias, which for most shells results in a significant speed
improvement.
.TP
.I "\-S"
Set colorization to
.I no
if the terminal does not occur in any
.B TERM
statement, even if there is a
.B COLOR
statement in the global section of the file.  This is for
compatibility with Slackware GNU
.B ls
for which the order of the statements did not matter.
.TP
.I "\-v"
Display a version string and exit.
.SS CONFIGURATION FILE
If a filename is specified on the command line, it will be used to
generate the setup, if not, the file
.B .dir_colors
in the users home directory will be used.  If that file does not exist
either, the system-wide file of
.B /etc/DIR_COLORS
will be used.
.PP
The configuration file consists of several statements, one per line.
Anything right of a hash mark (\#) is treated as a comment, if the
hash mark is at the beginning of a line or is preceded by at least one
whitespace.  Blank lines are ignored.
.PP
The
.I global
section of the file consists of any statement before the first
.B TERM
statement.  Any statement in the global section of the file is
considered valid for all terminal types  Following the global section
is one or more 
.I terminal-specific
sections, which are preceded by one or more
.B TERM
statements which specify the terminal types (as given by the
.B TERM
environment variable) the following declarations apply for.  It is
always possible to override a global declaration by a subsequent
terminal-specific one.
.PP
The following statements are recognized, case is insignificant:
.PP
.TP
.B TERM \fIterminal-type\fR
Starts a terminal-specific section and specifies which terminal it
applies to.  Multiple
.B TERM
statements can be used to create a section which applies for several
terminal types.
.TP
.B COLOR yes|all|no|none|tty
Specifies that colorization should be always enabled (\fIyes\fR or
\fIall\fR), never enabled (\fIno\fR or \fInone\fR) or enabled only if
the output is a terminal (\fItty\fR).  The default is \fIno\fR.
.TP
.B OPTIONS \fIoptions\fR
Adds command line options to the default
.B ls
command line.  The options can be any valid
.B ls
command line options, and should include the leading minus sign.
Please note that
.B dircolors
does not verify the validity of these options.
.TP
.B NORMAL \fIcolor-sequence\fR
Specifies the color used for normal (non-filename) text.
.TP
.B FILE \fIcolor-sequence\fR
Specifies the color used for a regular file.
.TP
.B DIR \fIcolor-sequence\fR
Specifies the color used for directories.
.TP
.B LINK \fIcolor-sequence\fR
Specifies the color used for a symbolic link.
.TP
.B ORPHAN \fIcolor-sequence\fR
Specifies the color used for an orphanned symbolic link (one which
points to a nonexistent file).  If this is unspecified,
.B ls
will use the
.B LINK
color instead.
.TP
.B MISSING \fIcolor-sequence\fR
Specifies the color used for a missing file (a nonexistent file which
nevertheless has a symbolic link pointing to it).  If this is unspecified,
.B ls
will use the
.B FILE
color instead.
.TP
.B FIFO \fIcolor-sequence\fR
Specifies the color used for a FIFO (named pipe).
.TP
.B SOCK \fIcolor-sequence\fR
Specifies the color used for a socket.
.TP
.B BLK \fIcolor-sequence\fR
Specifies the color used for a block device special file.
.TP
.B CHR \fIcolor-sequence\fR
Specifies the color used for a character device special file.
.TP
.B EXEC \fIcolor-sequence\fR
Specifies the color used for a file with the executable attribute set.
.TP
.B LEFTCODE \fIcolor-sequence\fR
Specifies the
.I "left code"
for non-ISO\ 6429 terminals (see below).
.TP
.B RIGHTCODE \fIcolor-sequence\fR
Specifies the
.I "right code"
for non-ISO\ 6429 terminals (see below).
.TP
.B ENDCODE \fIcolor-sequence\fR
Specifies the
.I "end code"
for non-ISO\ 6429 terminals (see below).
.TP
\fB*\fIextension\fR \fIcolor-sequence\fR
Specifies the color used for any file that ends in \fIextension\fR.
.TP
\fB .\fIextension\fR \fIcolor-sequence\fR
Same as \fB*\fR.\fIextension\fR.  Specifies the color used for any file that
ends in .\fIextension\fR.  Note that the period is included in the
extension, which makes it impossible to specify an extension not
starting with a period, such as
.B ~
for
.B emacs
backup files.  This form should be considered obsolete.
.SH ISO 6429 (ANSI) COLOR SEQUENCES
Most color-capable ASCII terminals today use ISO 6429 (ANSI) color sequences,
and many common terminals without color capability, including
.B xterm
and the widely used and cloned DEC VT100, will recognize ISO 6429 color
codes and harmlessly eliminate them from the output or emulate them.
.B ls
uses ISO 6429 codes by default, assuming colorization is enabled.

ISO 6429 color sequences are composed of sequences of numbers
separated by semicolons.  The most common codes are:
.sp
.RS +.2i
.ta 1.0i
.nf
 0	to restore default color
 1	for brighter colors 
 4	for underlined text
 5	for flashing text
30	for black foreground
31	for red foreground
32	for green foreground
33	for yellow (or brown) foreground
34	for blue foreground
35	for purple foreground
36	for cyan foreground
37	for white (or gray) foreground
40	for black background
41	for red background
42	for green background
43	for yellow (or brown) background
44	for blue background
45	for purple background
46	for cyan background
47	for white (or gray) background
.fi
.RE
.sp
Not all commands will work on all systems or display devices.
.PP
.B ls
uses the following defaults:
.sp
.RS +.2i
.ta 1.0i 2.5i
.nf
\fBNORMAL\fR	0	Normal (non-filename) text
\fBFILE\fR	0	Regular file
\fBDIR\fR	32	Directory
\fBLINK\fR	36	Symbolic link
\fBORPHAN\fR	undefined	Orphanned symbolic link
\fBMISSING\fR	undefined	Missing file
\fBFIFO\fR	31	Named pipe (FIFO)
\fBSOCK\fR	33	Socket
\fBBLK\fR	44;37	Block device
\fBCHR\fR	44;37	Character device
\fBEXEC\fR	35	Executable file
.fi
.RE
.sp
A few terminal programs do not recognize the default 
properly.  If all text gets colorized after you do a directory
listing, change the
.B NORMAL
and
.B FILE
codes to the numerical codes for your normal foreground and background
colors.
.SH OTHER TERMINAL TYPES (ADVANCED CONFIGURATION)
If you have a color-capable (or otherwise highlighting) terminal (or
printer!) which uses a different set of codes, you can still generate
a suitable setup.  To do so you will have to use the
.BR LEFTCODE ,
.BR RIGHTCODE ,
and
.BR ENDCODE
definitions.
.PP
When writing out a filename,
.B ls
generates the following output sequence:
.B LEFTCODE
.I typecode
.B RIGHTCODE
.I filename
.BR ENDCODE ,
where the
.I typecode
is the color sequence that depends on the type or name of file.  If the
.B ENDCODE
is undefined, the sequence
.B "LEFTCODE NORMAL RIGHTCODE"
will be used instead.  The purpose of the left- and rightcodes is
merely to reduce the amount of typing necessary (and to hide ugly
escape codes away from the user).  If they are not appropriate for
your terminal, you can eliminate them by specifying the respective
keyword on a line by itself.
.PP
.B NOTE:
If the
.B ENDCODE
is defined in the global section of the setup file, it
.I cannot
be undefined in a terminal-specific section of the file.  This means
any
.B NORMAL
definition will have no effect.  A different
.B ENDCODE
can however be specified, which would have the same effect.
.SH ESCAPE SEQUENCES
To specify control- or blank characters in the color sequences or
filename extensions, either C-style \e-escaped notation, or
.BR stty -style
^-notation can be used.  The C-style notation
includes the following characters:
.sp
.RS +.2i
.ta 1.0i
.nf
\fB\ea\fR	Bell (ASCII 7)
\fB\eb\fR	Backspace (ASCII 8)
\fB\ee\fR	Escape (ASCII 27)
\fB\ef\fR	Form feed (ASCII 12)
\fB\en\fR	Newline (ASCII 10)
\fB\er\fR	Carriage Return (ASCII 13)
\fB\et\fR	Tab (ASCII 9)
\fB\ev\fR	Vertical Tab (ASCII 11)
\fB\e?\fR	Delete (ASCII 127)
\fB\e\fInnn\fR	Any character (octal notation)
\fB\ex\fInnn\fR	Any character (hexadecimal notation)
\fB\e_\fR	Space
\fB\e\e\fR	Backslash (\e)
\fB\e^\fR	Caret (^)
\fB\e#\fR	Hash mark (#)
.fi
.RE
.sp
Please note that escapes are necessary to enter a space, backslash,
caret or any control character anywhere in the string, as well as a
hash mark as the first character.
.SH NOTES
The default
.B LEFTCODE
and
.B RIGHTCODE
definitions, which are used by ISO 6429 terminals are:
.sp
.RS +.2i
.ta 1.0i
.nf
\fBLEFTCODE\fR	\ee[
\fBRIGHTCODE\fR	m
.fi
.RE
.sp
The default
.B ENDCODE
is undefined.
.SH AUTHOR
H. Peter Anvin <Peter.Anvin@linux.org> with input from Patrick
Volkerding, creator of the Slackware Linux distribution.
.SH BUGS
Currently is somewhat poorly integrated into the GNU fileutils
package.  For example, it does not support long options and is
probably pickier about POSIX violations (i.e. less portable) 
than the rest of the programs.
.PP
The support for non-ISO 6429 terminals is a kluge at the very best.
.SH FILES
.ta \w'/etc/DIR_COLORS'u+3n
.nf
\fB/etc/DIR_COLORS\fR	System-wide setup file
\fB~/.dir_colors\fR	User setup file
.fi
.SH SEE ALSO
.BR ls (1),
.BR stty (1),
.BR xterm (1)


NEWS
Frugalware 0.1rc2 released
Mon Oct 11 15:40:13 CEST 2004
VMiklos
The Frugalware Developer Team is pleased to announce the immediate availability of Frugalware 0.1rc2, the second release candidate of the 0.1 release.
For a list of improvements and fixed bugs since rc1, please refer to the Frugalware <a href=changelog.php>Change Log</a>.
Frugalware 0.1rc2 can be downloaded by visiting the <a href=download.php>download</a> page.

Frugalware 0.1rc1 released
Wed Sep 29 07:14:26 CEST 2004
VMiklos
The Frugalware Developer Team is pleased to announce the immediate availability of Frugalware 0.1rc1, the first release candidate of the 0.1 release.
For a list of improvements since the beginning, please refer to the Frugalware <a href=changelog.php>Change Log</a>.
Frugalware 0.1rc1 can be downloaded by visiting the <a href=download.php>download</a> page.

Test message
Mon Sep 13 13:53:37 CEST 2004
VMiklos
seems to work... ;-)


source/base/coreutils/dircolors.c
/*
   dircolors.c

   Parse a Slackware-style DIR_COLORS file

   Copyright (C) 1994, 1995 H. Peter Anvin

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <string.h>

#define ETC_DIR "/etc"
#define USER_FILE ".dir_colors"	/* Versus user's home directory */
#define SYSTEM_FILE "/DIR_COLORS" /* System-wide file in directory SYSTEM_DIR
				    (defined on the cc command line) */

#define STRINGLEN 2048		/* Max length of a string */

enum modes { mo_sh, mo_csh, mo_ksh, mo_zsh, mo_unknown, mo_err };

const char *shells[] = { "sh", "ash", "csh", "tcsh", "bash", "ksh",
			 "zsh", NULL };

const int shell_mode[] = { mo_sh, mo_sh, mo_csh, mo_csh,
			   mo_ksh, mo_ksh, mo_zsh };

static int
figure_mode (me)
     char *me;
{
  char *shell, *shellv;
  int i;

  shellv = getenv("SHELL");
  if ( !shellv || !(*shellv) )
    {
      fprintf(stderr, "%s: No SHELL variable, and no mode option specified\n",
	      me);
      return mo_err;
    }

  if ( (shell = strrchr(shellv,'/')) != NULL )
    shell++;
  else
    shell = shellv;

  for ( i = 0 ; shells[i] ; i++ )
    {
      if ( strcmp(shell,shells[i]) == 0 )
	return shell_mode[i];
    }

  fprintf(stderr, "%s: Unknown shell `%s'\n", me, shell);
  return mo_err;
}

static void
parse_line(char **keyword, char **arg, char *line)
{
  char *p;

  *keyword = *arg = "";

  for ( p = line ; isspace(*p) ; p++ );

  if ( ! (*p) || *p == '#' )
    return;

  *keyword = p;

  for ( ; !isspace(*p) ; p++ )
    {
      if ( !(*p) )
	return;
    }

  *(p++) = '\0';

  for ( ; isspace(*p) ; p++ );
  
  if ( ! (*p) || *p == '#' )
    return;
  
  *arg = p;

  for ( ; *p != '\0' && *p != '#' ; p++ );
  for ( p-- ; isspace(*p) ; p-- );
  p++;

  *p = '\0';

  return;
}

/* Write a string to standard out, while watching for "dangerous"
   sequences like unescaped : and = characters */

static void
put_seq(str, follow)
     char *str;
     char follow;
{
  int danger = 1;

  for ( ; *str ; str++ )
    {
      switch ( *str )
	{
	case '\\':
	case '^':
	  danger = !danger;
	  break;

	case ':':
	case '=':
	  if ( danger )
	    putchar('\\');
	  /* Fall through */

	default:
	  danger = 1;
	  break;
	}

      putchar( *str );
    }

  putchar(follow);		/* The character that ends the sequence */
}    

/* Parser needs these state variables */
enum states { st_termno, st_termyes, st_termsure, st_global };

const char *slack_codes[] = {"NORMAL", "NORM", "FILE", "DIR", "LNK", "LINK",
"SYMLINK", "ORPHAN", "MISSING", "FIFO", "PIPE", "SOCK", "BLK", "BLOCK",
"CHR", "CHAR", "EXEC", "LEFT", "LEFTCODE", "RIGHT", "RIGHTCODE", "END",
"ENDCODE", NULL};

const char *ls_codes[] = {"no", "no", "fi", "di", "ln", "ln", "ln",
"or", "mi", "pi", "pi", "so", "bd", "bd", "cd", "cd", "ex", "lc", "lc", "rc",
"rc", "ec", "ec"};

enum color_opts { col_yes, col_no, col_tty };

int
main (argc, argv)
     int argc;
     char *argv[];
{
  char *p, *q;
  char *file = NULL;
  int i;
  int mode = mo_unknown;
  FILE *fp = NULL;
  char *term;
  int state;

  char line[STRINGLEN];
  char useropts[2048] = "";
  char *keywd, *arg;

  int eightbit = 1;		/* Default to 8-bit */
  int color_opt = col_no;	/* Assume --color=no */

  int strict_slack = 0;		/* Strict Slackware compatibility */
  int no_path = 0;		/* Do not search PATH */

  char *copt, *bopt;
  
  /* Parse command line */

  for ( i = 1 ; i < argc ; i++ )
    {
      if ( argv[i][0] == '-' )
	{
	  for ( p = &argv[i][1] ; *p ; p++ )
	    {
	      switch ( *p )
		{
		case 'a':
		case 's':	/* Plain sh mode */
		  mode = mo_sh;
		  break;

		case 'c':
		case 't':
		  mode = mo_csh;
		  break;

		case 'b':
		case 'k':
		  mode = mo_ksh;
		  break;

		case 'z':
		  mode = mo_zsh;
		  break;

		case 'P':
		  no_path = 1;
		  break;

		case 'S':
		  strict_slack = 1;
		  eightbit = 0;	/* Default to 7-bit */
		  break;

		case 'v':
		  printf("For %s with color-ls patch %s\n",
			 "fileutils-3.16", "3.12.0.3");
		  exit(0);

		default:
		  fprintf(stderr, "%s: Unknown option -%c\n", argv[0], *p);
		  exit(1);
		}
	    }
	}
      else
	file = argv[i];
    }

  /* Use shell to determine mode, if not already done. */

  if ( mode == mo_unknown )
    {
      mode = figure_mode(argv[0]);
      if ( mode == mo_err )
	exit(1);
    }

  /* Open dir_colors file */

  if ( !file )
    {
      p = getenv("HOME");
      if ( p && *p )
	{
	  chdir(p);
	  fp = fopen(USER_FILE, "r");
	}

      if ( !fp )
	fp = fopen(ETC_DIR SYSTEM_FILE, "r");
    }
  else
    fp = fopen(file, "r");
  
  if ( !fp )
    {
      perror(argv[0]);
      exit(1);
    }

  /* Get terminal type */

  term = getenv("TERM");
  if ( !term || !(*term) )
    term = "none";

  /* Write out common start */

  switch ( mode )
    {
    case mo_csh:
      printf("set noglob;\n\
setenv LS_COLORS \':");
      break;
    case mo_sh:
    case mo_ksh:
    case mo_zsh:
      printf("LS_COLORS=\'");
      break;
    }

  /* Start parsing that sucker */

  state = st_global;

  while ( fgets(line,STRINGLEN,fp) != NULL )
    {
      parse_line(&keywd, &arg, line);
      if ( *keywd )
	{
	  if ( strcasecmp(keywd, "TERM") == 0 )
	    {
	      if ( strcmp(arg, term) == 0 )
		{
		  state = st_termsure;
		  strict_slack = 0; /* We've fulfilled the requirement */
		}
	      else if ( state != st_termsure )
		state = st_termno;
	    }
	  else
	    {
	      if ( state == st_termsure )
		state = st_termyes; /* Another TERM can cancel */
	      
	      if ( state != st_termno )
		{
		  if ( keywd[0] == '.' )
		    {
		      putchar('*');
		      put_seq(keywd,'=');
		      put_seq(arg,':');
		    }
		  else if ( keywd[0] == '*' )
		    {
		      put_seq(keywd,'=');
		      put_seq(arg,':');
		    }
		  else if ( strcasecmp(keywd, "OPTIONS") == 0 )
		    {
		      strcat(useropts, " ");
		      strcat(useropts, arg);
		    }
		  else if ( strcasecmp(keywd, "COLOR") == 0 )
		    {
		      switch ( arg[0] )
			{
			case 'a':
			case 'y':
			case '1':
			  color_opt = col_yes;
			  break;

			case 'n':
			case '0':
			  color_opt = col_no;
			  break;

			case 't':
			  color_opt = col_tty;
			  break;
			  
			default:
			  fprintf(stderr, "%s: Unknown COLOR option `%s'\n",
				  argv[0], arg);
			  break;
			}
		    }
		  else if ( strcasecmp(keywd, "EIGHTBIT") == 0 )
		    {
		      switch( arg[0] )
			{
			case 'y':
			case '1':
			  eightbit = 1;
			      break;
			  
			case 'n':
			case '0':
			  eightbit = 0;
			  break;
			  
			default:
			  fprintf(stderr, "%s: Unknown EIGHTBIT option `%s'\n",
				  argv[0], arg);
			  break;
			}
		    }
		  else
		    {
		      for ( i = 0 ; slack_codes[i] ; i++ )
			{
			  if ( strcasecmp(keywd, slack_codes[i]) == 0 )
			    break;
			}

		      if ( slack_codes[i] )
			{
			  printf("%s=", ls_codes[i]);
			  put_seq(arg,':');
			}
		      else
			fprintf(stderr, "%s: Unknown keyword %s\n",
				argv[0], keywd);
		    }
		}
	    }
	}
    }

  fclose(fp);

  /* If strict_slack is still set, we force COLOR to no */

  if ( strict_slack )
    color_opt = col_no;

  /* Decide on the options */
  
  switch ( color_opt )
    {
    case col_yes:
      copt = "--color=always";
      break;

    case col_no:
      copt = "--color=never";
      break;

    case col_tty:
      copt = "--color=auto";
      break;
    }

  /* bopt = eightbit ? "--8bit" : "--7bit"; */
  bopt = ""; /* obsolete; cruft */

  /* Find ls in the path */
  
  if ( !no_path )
    {
      no_path = 1;		/* Assume we won't find one */

      p = getenv("PATH");
      if ( p && *p )
	{
	  while ( *p )
	    {
	      while ( *p == ':' )
		p++;

	      if ( *p != '/' )	/* Skip relative path entries */
		while ( *p && *p != ':' )
		  p++;
	      else
		{
		  q = line;
		  while ( *p && *p != ':' )
		    *(q++) = *(p++);
		  /* Make sure it ends in slash */
		  if ( *(q-1) != '/' )
		    *(q++) = '/';

		  strcpy(q, "ls");
		  if ( access(line, X_OK) == 0 )
		    {
		      no_path = 0; /* Found it */
		      break;
		    }
		}
	    }
	}
    }
  
  /* Write it out */

  switch ( mode )
    {
    case mo_sh:
      if ( no_path )
	printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS='%s %s%s';\n\
export LS_OPTIONS;\n\
ls () { ( exec ls $LS_OPTIONS \"$@\" ) };\n\
dir () { ( exec dir $LS_OPTIONS \"$@\" ) };\n\
vdir () { ( exec vdir $LS_OPTIONS \"$@\" ) };\n\
d () { dir \"$@\" ; };\n\
v () { vdir \"$@\" ; };\n", bopt, copt, useropts);
      else
	printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS='%s %s%s';\n\
ls () { %s $LS_OPTIONS \"$@\" ; };\n\
dir () { %s $LS_OPTIONS --format=vertical \"$@\" ; };\n\
vdir () { %s $LS_OPTIONS --format=long \"$@\" ; };\n\
d () { dir \"$@\" ; };\n\
v () { vdir \"$@\" ; };\n", bopt, copt, useropts, line, line, line);
      break;

    case mo_csh:
      if ( no_path )
	printf("\';\n\
setenv LS_OPTIONS '%s %s%s';\n\
alias ls \'ls $LS_OPTIONS\';\n\
alias dir \'dir $LS_OPTIONS\';\n\
alias vdir \'vdir $LS_OPTIONS\';\n\
alias d dir;\n\
alias v vdir;\n\
unset noglob;\n", bopt, copt, useropts);
      else
	printf("\';\n\
setenv LS_OPTIONS '%s %s%s';\n\
alias ls \'%s $LS_OPTIONS\';\n\
alias dir \'%s $LS_OPTIONS --format=vertical\';\n\
alias vdir \'%s $LS_OPTIONS --format=long\';\n\
alias d dir;\n\
alias v vdir;\n\
unset noglob;\n", bopt, copt, useropts, line, line, line);
      break;

    case mo_ksh:
      if ( no_path )
      printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS='%s %s%s';\n\
export LS_OPTIONS;\n\
alias ls=\'ls $LS_OPTIONS\';\n\
alias dir=\'dir $LS_OPTIONS\';\n\
alias vdir=\'vdir $LS_OPTIONS\';\n\
alias d=dir;\n\
alias v=vdir;\n", bopt, copt, useropts);
      else
	printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS='%s %s%s';\n\
export LS_OPTIONS;\n\
alias ls=\'%s $LS_OPTIONS\';\n\
alias dir=\'%s $LS_OPTIONS --format=vertical\';\n\
alias vdir=\'%s $LS_OPTIONS --format=long\';\n\
alias d=dir;\n\
alias v=vdir;\n", bopt, copt, useropts, line, line, line);
      break;

    case mo_zsh:
      if ( no_path )
      printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS=(%s %s%s);\n\
export LS_OPTIONS;\n\
alias ls=\'ls $LS_OPTIONS\';\n\
alias dir=\'dir $LS_OPTIONS\';\n\
alias vdir=\'vdir $LS_OPTIONS\';\n\
alias d=dir;\n\
alias v=vdir;\n", bopt, copt, useropts);
      else
	printf("\';\n\
export LS_COLORS;\n\
LS_OPTIONS=(%s %s%s);\n\
export LS_OPTIONS;\n\
alias ls=\'%s $LS_OPTIONS\';\n\
alias dir=\'%s $LS_OPTIONS --format=vertical\';\n\
alias vdir=\'%s $LS_OPTIONS --format=long\';\n\
alias d=dir;\n\
alias v=vdir;\n", bopt, copt, useropts, line, line, line);
      break;
    }

  exit(0);
}
